### DISCUSSION TITLE
HRM not working on fresh installation on a Pi Zero 2

### DISCUSSION THREAD

**Discussion by Gordon-Shumway2** [JaapvanEkris/openrowingmonitor#35] [General]
I'm afraid I am seeing the same error stack that I mentioned in the Pi Zero (1) discussion with a fresh OS installation on the latest v1beta_updates branch using  a Pi Zero 2. Heart rate monitor (TickrX 2) is not recognized and clicking on the BLE icon throws this error stack: 

> Feb 05 22:03:40 pirower npm[1169]: webserver running on port 80
Feb 05 22:03:41 pirower npm[1188]: /opt/openrowingmonitor/node_modules/@abandonware/noble/lib/hci-socket/bindings.js:16
Feb 05 22:03:41 pirower npm[1188]:   this._isExtended = 'extended' in options && options.extended;
Feb 05 22:03:41 pirower npm[1188]:                                 ^
Feb 05 22:03:41 pirower npm[1188]: TypeError: Cannot use 'in' operator to search for 'extended' in undefined
Feb 05 22:03:41 pirower npm[1188]:     at new NobleBindings (/opt/openrowingmonitor/node_modules/@abandonware/noble/lib/hci-socket/bindings.js:16:33)
Feb 05 22:03:41 pirower npm[1188]:     at file:///opt/openrowingmonitor/app/peripherals/ble/hrm/HeartRateManager.js:49:25
Feb 05 22:03:41 pirower npm[1188]:     at ModuleJob.run (node:internal/modules/esm/module_job:195:25)
Feb 05 22:03:41 pirower npm[1188]:     at async ModuleLoader.import (node:internal/modules/esm/loader:336:24)
Feb 05 22:03:41 pirower npm[1188]:     at async loadESM (node:internal/process/esm_loader:34:7)
Feb 05 22:03:41 pirower npm[1188]:     at async handleMainPromise (node:internal/modules/run_main:106:12)
Feb 05 22:03:41 pirower npm[1188]: Node.js v18.19.0

Clicking on the ANT icon (without any ANT hardware being present) leads to ORM exiting: 

> Feb 05 21:58:07 pirower npm[951]: bluetooth profile: FTMS Indoor Bike
Feb 05 21:58:09 pirower npm[951]: bluetooth profile: Concept2 PM5
Feb 05 21:58:10 pirower npm[951]: bluetooth profile: Cycling Speed and Cadence
Feb 05 21:58:11 pirower npm[951]: bluetooth profile: Cycling Power Meter
Feb 05 21:58:12 pirower npm[951]: bluetooth profile: Off
Feb 05 21:58:14 pirower npm[951]: bluetooth profile: FTMS Rower
Feb 05 21:58:18 pirower npm[951]: ant plus profile: FE
Feb 05 21:58:18 pirower npm[951]: Error: Error opening Ant Stick
Feb 05 21:58:18 pirower npm[951]:     at AntManager.openAntStick (file:///opt/openrowingmonitor/app/peripherals/ant/AntManager.js:21:47)
Feb 05 21:58:18 pirower npm[951]:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Feb 05 21:58:18 pirower npm[951]:     at async Object.attach (file:///opt/openrowingmonitor/app/peripherals/ant/FEPeripheral.js:41:38)
Feb 05 21:58:18 pirower npm[951]:     at async createAntPeripheral (file:///opt/openrowingmonitor/app/peripherals/PeripheralManager.js:178:11)
Feb 05 21:58:24 pirower npm[951]: ANT+ FE server stopped [deviceId=7687 channel=1]
Feb 05 21:58:24 pirower npm[951]: Uncaught Exception: TypeError: Cannot read properties of undefined (reading 'transfer')
Feb 05 21:58:24 pirower npm[951]:     at AntDevice.write (/opt/openrowingmonitor/node_modules/incyclist-ant-plus/lib/ant-device.js:296:20)
Feb 05 21:58:24 pirower npm[951]:     at file:///opt/openrowingmonitor/app/peripherals/ant/FEPeripheral.js:69:18
Feb 05 21:58:24 pirower npm[951]:     at new Promise (<anonymous>)
Feb 05 21:58:24 pirower npm[951]:     at Object.destroy (file:///opt/openrowingmonitor/app/peripherals/ant/FEPeripheral.js:60:12)
Feb 05 21:58:24 pirower npm[951]:     at createAntPeripheral (file:///opt/openrowingmonitor/app/peripherals/PeripheralManager.js:157:28)
Feb 05 21:58:24 pirower npm[951]:     at EventEmitter.switchAntPeripheralMode (file:///opt/openrowingmonitor/app/peripherals/PeripheralManager.js:151:5)
Feb 05 21:58:24 pirower npm[951]:     at EventEmitter.<anonymous> (file:///opt/openrowingmonitor/app/server.js:243:25)
Feb 05 21:58:24 pirower npm[951]:     at EventEmitter.emit (node:events:517:28)
Feb 05 21:58:24 pirower npm[951]:     at WebSocket.incoming (file:///opt/openrowingmonitor/app/WebServer.js:38:19)
Feb 05 21:58:24 pirower npm[951]:     at WebSocket.emit (node:events:517:28)
Feb 05 21:58:24 pirower npm[951]: shutting down all peripherals
Feb 05 21:58:24 pirower npm[951]: ANT+ FE server stopped [deviceId=7687 channel=1]
Feb 05 21:58:24 pirower npm[951]: peripheral shutdown was unsuccessful, restart of Pi may required TypeError: Cannot read properties of undefined (reading 'transfer')
Feb 05 21:58:24 pirower npm[951]:     at AntDevice.write (/opt/openrowingmonitor/node_modules/incyclist-ant-plus/lib/ant-device.js:296:20)
Feb 05 21:58:24 pirower npm[951]:     at file:///opt/openrowingmonitor/app/peripherals/ant/FEPeripheral.js:69:18
Feb 05 21:58:24 pirower npm[951]:     at new Promise (<anonymous>)
Feb 05 21:58:24 pirower npm[951]:     at Object.destroy (file:///opt/openrowingmonitor/app/peripherals/ant/FEPeripheral.js:60:12)
Feb 05 21:58:24 pirower npm[951]:     at EventEmitter.shutdownAllPeripherals (file:///opt/openrowingmonitor/app/peripherals/PeripheralManager.js:272:28)
Feb 05 21:58:24 pirower npm[951]:     at async process.<anonymous> (file:///opt/openrowingmonitor/app/server.js:156:3)
Feb 05 21:58:25 pirower systemd[1]: openrowingmonitor.service: Main process exited, code=exited, status=1/FAILURE
Feb 05 21:58:25 pirower npm[965]: 2024-02-05 21:58:25 sigHandler: Unhandled signal 18, terminating
Feb 05 21:58:25 pirower systemd[1]: openrowingmonitor.service: Failed with result 'exit-code'.
Feb 05 21:58:25 pirower systemd[1]: openrowingmonitor.service: Consumed 19.509s CPU time.
Feb 05 21:58:25 pirower systemd[1]: openrowingmonitor.service: Scheduled restart job, restart counter is at 1.
Feb 05 21:58:25 pirower systemd[1]: Stopped Open Rowing Monitor.
>

Is this a known issue? 
I used parts of my old config (rower-specific things, not BLE or ANT related) in both installations so this might well be a misconfiguration issue but so far I could not find anything obviously wrong. 

Kind regards, 
Michael

**Comment by JaapvanEkris** [comment]
Hi Michael,

Thanks for catching this and reporting it! I can reproduce this as well. I am going to downgrade the bluethooth libraries again. This is weird, as I have tested it earlier in the process.

Odd thing is, when I downgrade the libraries manually, it still won't work. I'm now seeing what happens when I revert to the complete original setup.

Jaap

**Reply by JaapvanEkris** [reply]
When I restore all old packages (yes, it is drastic), then it works. I'll investigate further tomorrow, and see who the culprit really is.

A sidenote: broadcasting my HR from my Garmin Epix actually crashes BleNo as some value is out of bounds.

Time to sleep.

**Comment by JaapvanEkris** [comment]
I've got HRM working again on BullsEye with Node18. I'l test further with BullsEye on Node 20, and Bookworm on Node20.

**Comment by JaapvanEkris** [comment]
After some initial tests, it all seems to work. You can now install on bullseye and bookworm. HRM works, as does EXR.

**Comment by Gordon-Shumway2** [comment]
I can confirm that ORM and HRM with a TickrX 2 now works on my Pi Zero 2 after running "/opt/openrowingmonitor/bin/updateopenrowingmonitor.sh". 

Thank you for the quick fix!
Best regards, 
Michael
